### PR TITLE
Abfs blob endpoint support common code list blob: WIP

### PR DESCRIPTION
**Pull request by saxenapranav**
bloblist api support

https://learn.microsoft.com/en-us/rest/api/storageservices/list-blobs?tabs=azure-ad

### CODE REVIEW

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 1]
```diff
@@ -497,10 +500,39 @@ public BlobProperty getBlobProperty(Path blobPath, TracingContext tracingContext
...
+   *
+   * @param sourceDirBlobPath path from where the list of blob is requried.
+   * @param tracingContext object of {@link TracingContext}
+   * @param maxResult define how many blobs can client handle in server response.
+   * In case maxResult <= 5000, server sends number of blobs equal to the value. In
+   * case maxResult > 5000, server sends maximum 5000 blobs.
+   * @return List of blobProperties
+   * @throws AbfsRestOperationException exception from server-calls / xml-parsing
+   */
+  public List<BlobProperty> getDirectoryBlobProperty(Path sourceDirBlobPath,
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Can we change the name to something like getListBlobs why are we specifically mentioning directory here ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsHttpOperation.java [hunk 2]
```diff
@@ -414,7 +425,11 @@ public void processResponse(final byte[] buffer, final int offset, final int len
           if (url.toString().contains(COMP_BLOCKLIST)){
             parseBlockListResponse(stream);
           } else {
-            parseListFilesResponse(stream);
+            if(url.toString().contains(COMP_LIST)) {
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsHttpOperation.java]: nit: space after if

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobListXmlParser.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    this.blobList = blobList;
+  }
+
+  @Override
+  public void startElement(final String uri,
+      final String localName,
+      final String qName,
+      final Attributes attributes) throws SAXException {
+    elements.push(localName);
+    if(AbfsHttpConstants.BLOB.equals(localName)) {
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobListXmlParser.java]: Spaces not added after if, will lead to checkstyle errors

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobListXmlParser.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      currentBlobProperty = new BlobProperty();
+    }
+  }
+
+  @Override
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobListXmlParser.java:43]: Can you please add an example of a xml response to understand the xml parsing better ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobProperty.java [hunk 5]
```diff
@@ -1,29 +1,49 @@
...
   private Path path;
   private String url;
 
   private String copySourceUrl;
   private String copyId;
   private String copyStatus;
   private String statusDescription;
-  private int contentLength = 0;
+  private Long contentLength = 0L;
+
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobProperty.java]: Additional spaces can be removed.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/constants/HttpQueryParams.java [hunk 6]
```diff
@@ -29,6 +29,10 @@ public final class HttpQueryParams {
   public static final String QUERY_PARAM_RESOURCE = "resource";
   public static final String QUERY_PARAM_RESTYPE = "restype";
   public static final String QUERY_PARAM_COMP = "comp";
+  public static final String LIST = "list";
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/constants/HttpQueryParams.java]: Should this be QUERY_PARAM_LIST ?

Comment (saxenapranav) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/constants/HttpQueryParams.java]: Have refactored to QUERY_PARAM_COMP_VALUE_LIST

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobListXmlParser.java (hunk 4 above)
Comment (saxenapranav) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobListXmlParser.java:43]: Added example link in javadocs for the BlobListXmlParser class

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 7]
```diff
@@ -497,10 +498,39 @@ public BlobProperty getBlobProperty(Path blobPath, TracingContext tracingContext
     blobProperty.setCopySourceUrl(opResult.getResponseHeader(X_MS_COPY_SOURCE));
     blobProperty.setStatusDescription(opResult.getResponseHeader(X_MS_COPY_STATUS_DESCRIPTION));
     blobProperty.setCopyStatus(opResult.getResponseHeader(X_MS_COPY_STATUS));
-    blobProperty.setContentLength(Integer.parseInt(opResult.getResponseHeader(CONTENT_LENGTH)));
+    blobProperty.setContentLength(Long.parseLong(opResult.getResponseHeader(CONTENT_LENGTH)));
     return blobProperty;
   }
 
+  /**
+   * Call server API for ListBlob on the blob endpoint. This API return a limited
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: nit: returns*

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java [hunk 8]
```diff
@@ -1137,8 +1137,52 @@ public AbfsRestOperation getBlobProperty(Path blobPath, TracingContext tracingCo
...
+   * @return list of {@link BlobProperty}
+   * @throws AzureBlobFileSystemException thrown from server-call / xml-parsing
+   */
+  public AbfsRestOperation getListBlobs(Path sourceDirBlobPath,
+      TracingContext tracingContext, String marker, String prefix, Integer maxResult)
+      throws AzureBlobFileSystemException {
+    AbfsUriQueryBuilder abfsUriQueryBuilder = createDefaultUriQueryBuilder();
+    abfsUriQueryBuilder.addQuery(QUERY_PARAM_RESTYPE, CONTAINER);
+    abfsUriQueryBuilder.addQuery(QUERY_PARAM_COMP, QUERY_PARAM_COMP_VALUE_LIST);
+    if(prefix == null) {
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java]: if, space

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java [hunk 9]
```diff
@@ -1137,8 +1137,52 @@ public AbfsRestOperation getBlobProperty(Path blobPath, TracingContext tracingCo
...
+    abfsUriQueryBuilder.addQuery(QUERY_PARAM_PREFIX, prefix);
+    if(marker != null) {
+      abfsUriQueryBuilder.addQuery(QUERY_PARAM_MARKER, marker);
+    }
+    if(maxResult != null) {
+      abfsUriQueryBuilder.addQuery(QUERY_PARAM_MAXRESULT, maxResult + "");
+    }
+    URL url = createRequestUrl(abfsUriQueryBuilder.toString());
+    try {
+      url = new URL(url.toString().replace("dfs.core", "blob.core"));
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java]: Should not  be hardcoded, use existing constants for prefix

Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java]: getListBlobs should only be called when the url already has blob endpoint in it, why do we need to do this replace ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobListXmlParser.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    }
+    if (parentNode.equals(AbfsHttpConstants.METADATA)) {
+      currentBlobProperty.addMetadata(currentNode, value);
+    }
+    if (parentNode.equals(AbfsHttpConstants.PROPERTIES)) {
+      if (currentNode.equals(AbfsHttpConstants.CONTENT_LEN)) {
+        currentBlobProperty.setContentLength(Long.valueOf(value));
+      }
+      if (currentNode.equals(AbfsHttpConstants.RESOURCE_TYPE)) {
+        if ("directory".equals(value)) {
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/BlobListXmlParser.java]: Should be a CONSTANT

**Review by anmolanmol1234 — APPROVED** [review]

**Pull request merged**